File: ucoder/1116-linhas-cruzadas.py
def ft_update(i,n):
    global ft_data
    while i <= n:
        ft_data[i] += 1
        i += i & (-i)

def ft_query(i):
    global ft_data
    answer = 0
    while i > 0:
        answer += ft_data[i]
        i -= i & (-i)
    return answer

n = int(input())
data = [int(x) for x in input().split(' ')]
ft_data = [0 for i in range(n+1)]

answer = 0
for i in range(n-1,-1,-1):
    answer += ft_query(data[i]-1)
    ft_update(data[i],n)
print(answer)

File: ucoder/1116-linhas-cruzadas.c
#include <stdio.h>

#ifdef WIN32
#define getchar_unlocked _getchar_nolock
#endif

#define MAX 60000

int ft_data[(MAX + 1)];

void ft_update(int iI, int iN){

    for(; iI <= iN; iI += (iI & (-iI)))
        ft_data[iI]++;

}

int ft_query(int iI){

    int iAnswer = 0;

    for(; iI > 0; iI -= (iI & (-iI)))
        iAnswer += ft_data[iI];

    return iAnswer;
        
}

void fastscan(int *x){

    register int c = 0;
    
    *x = 0;

    for(;c < '0' || c > '9'; c = getchar_unlocked());

    for(;c >= '0' && c <= '9'; c = getchar_unlocked()){
        *x = ((*x) << 1) + ((*x) << 3) + (c - '0');
    } 

}

int main(void){

    int iN, iI;
    int iData[MAX];
    int iAnswer = 0;

    fastscan(&iN);
    for(iI = 0; iI < iN; iI++)
        fastscan(&iData[iI]);

    for(iI = iN - 1; iI >= 0; iI--){
        iAnswer += ft_query(iData[iI] - 1);
        ft_update(iData[iI], iN);
    }

    printf("%d\n", iAnswer);

    return 0;
}